observatory-platform/observatory/platform/api/app/api/query.py

```diff
@@ -0,0 +1,316 @@
+"""
+example queries:
+http://0.0.0.0:5000/query?agg=institution&subset=journals&from=2018-01-15&to=2019-01-01&journal=Molecular
+%20Pharmaceutics&api_key=158f9f48b674b0fdd18e5b465d7ae37e 447 records, ~2 seconds
+http://0.0.0.0:5000/query?agg=country&subset=oa-metrics&api_key=158f9f48b674b0fdd18e5b465d7ae37e 11435 records, ~6 sec
+http://0.0.0.0:5000/query?agg=institution&subset=journals&id=grid.4691.a,
+grid.469280.1&api_key=158f9f48b674b0fdd18e5b465d7ae37e 28479 records, ~9 sec
+http://0.0.0.0:5000/query?agg=country&subset=collaborations&api_key=158f9f48b674b0fdd18e5b465d7ae37e 268162 records,
+~80 sec
+"""
+import os
```

Comment:
yes, I second this. I imagine there will soon be a fair few different use cases coming in for APIs. Particularly as we think about moving code to the oaebu repo. I could imagine there will be utiltiies, and core APIs that live in the observatory platform, some user-facing APIs that live in an OAEBU repo, and possibly some academic-observatory specific APIs that live in that repo too

Reply:
I think adding the Elasticsearch related code into its own file is a good idea. Happy to rename server.py -> app.py and query.py -> elastic.py.

However, it might be good to have a separate directory for these files, away from the Dockerfile and requirements.txt.
If these are all in the same directory it means that the complete image will have to be rebuild when something changes in the Dockerfile (or any of the other files in the directory, but this is the intent), because it copies itself into the image. If we have a separate directory we can avoid this as well as having a distinction between the actual API and the files necessary to deploy the API to google cloud.

Reply:
It shouldn't matter should it? If any file is changed you have to rebuild the image anyway?

Reply:
Ah, yes if you build and submit it to google cloud then you will have to rebuild the image. But for local development if you just build the Dockerfile after making changes in e.g. app.py you could use the cache up until the second last line.

It's quite a small file so it does not matter too much, but I find that installing the requirements can sometimes be a bit lengthy and annoying.